STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/az_ulib_registry.c

```diff
@@ -0,0 +1,300 @@
+    az_ulib_registry_node* runner;
+
+    /* Find occupied node that has maximum address such that a new key-value pair can be stored */
+    runner = (az_ulib_registry_node*)(uint8_t*)(&__REGISTRYINFO);
+    for (int i = 0; i < MAX_AZ_ULIB_REGISTRY_ENTRIES; i++)
+    {
+      if (runner->_internal.status.ready_flag == REGISTRY_READY)
+      {
+        if (az_span_ptr(runner->_internal.data.value) > max_address)
+        {
+          max_address = az_span_ptr(runner->_internal.data.value);
+          last_node = runner;
+        }
+      }
+      runner++;
+    }
+
+    /* Last_node should have reference to the last node in available memory, with rounded address */
+    int32_t value_buf_size;
+    value_buf_size = ((((az_span_size(last_node->_internal.data.value) - 1) >> 3) + 1) << 3);
+
+    /* Calculate final flash_ptr */
+    flash_ptr = az_span_ptr(last_node->_internal.data.value) + value_buf_size;
+  }
+
+  return flash_ptr;
+}
+
+AZ_NODISCARD az_result az_ulib_registry_init(void)
+{
+  /* Initialize lock */
+  az_pal_os_lock_init(&registry_lock);
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result az_ulib_registry_deinit(void)
+{
+  /* Deinitialize lock */
+  az_pal_os_lock_deinit(&registry_lock);
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result az_ulib_registry_delete(az_span key)
+{
+  /* Precondition check */
+  _az_PRECONDITION_VALID_SPAN(key, 1, false);
+
+  az_result result;
+  az_ulib_registry_node* runner;
+
+  /* Loop through registry for entry that matches the key */
+  runner = (az_ulib_registry_node*)(&__REGISTRYINFO);
+  for (int i = 0; i < MAX_AZ_ULIB_REGISTRY_ENTRIES; i++)
+  {
+    if (runner->_internal.status.ready_flag == REGISTRY_READY)
+    {
+      if (az_span_is_content_equal(key, runner->_internal.data.key))
+      {
+        set_registry_node_delete_flag((uint8_t*)runner);
+        result = AZ_OK;
+        break;
+      }
+    }
+    else
+    {
+      result = AZ_ERROR_ITEM_NOT_FOUND;
+      break;
+    }
+    runner++;
+  }
+  return result;
+}
+
+AZ_NODISCARD az_result az_ulib_registry_get(az_span key, az_span* value)
+{
+  /* Precondition check */
+  _az_PRECONDITION_VALID_SPAN(key, 1, false);
+
+  az_result result;
+  az_ulib_registry_node* runner;
+
+  /* Find node containing key */
+  runner = (az_ulib_registry_node*)(&__REGISTRYINFO);
+  for (int i = 0; i < MAX_AZ_ULIB_REGISTRY_ENTRIES; i++)
+  {
+    if(runner->_internal.status.delete_flag != REGISTRY_DELETED)
+    {
+      /* If the content is not deleted and ready for use */
+      if (runner->_internal.status.ready_flag == REGISTRY_READY)
+      {
+        if (az_span_is_content_equal(key, runner->_internal.data.key))
+        {
+          *value = runner->_internal.data.value;
+          result = AZ_OK;
+          break;
+        }
+      }
+      else 
+      {
+        result = AZ_ERROR_ITEM_NOT_FOUND;
+        break;
+      }
+    }
+    runner++;
+  }
+
+  return result;
+}
+
+AZ_NODISCARD az_result az_ulib_registry_add(az_span key, az_span value)
```

Comment:
>az_ulib_registry_add [](http://example.com/codeflow?start=23&length=20)

We are creating a technical debit here. The device can run out of memory because of fragmentation and the write only process. Please add a Task in DevOps to fix it.

Reply:
Created new feature in devops to track future work: #10617124